lib/membrane_opus/parser.ex

```diff
@@ -1,79 +1,254 @@
-  needed to separate the packet from byte stream and decode it. This parser drops the last frame size
-  from each packet header, converting packets to the basic version. The remaining frame size needs
-  to be derived from entire packet size.
-  See https://tools.ietf.org/html/rfc6716#appendix-B for details.
+  Adds the following metadata:
+
+  frame_size :: 2.5 | 5 | 10 | 20 | 40 | 60
+    Number of milliseconds per frame encoded in this packet.
+
+  frame_lengths :: [non_neg_integer()]
+    Ordered list of bytes used to encode each frame in this packet.
+    The length of this list is also the number of frames encoded in this packet.
+    However, a value of 0 indicates a dropped frame and should be accounted
+    for when using this field.
   """
+
   use Membrane.Filter
 
-  alias Membrane.{Buffer, Opus, RemoteStream}
-  alias Membrane.Opus.PacketUtils
+  alias Membrane.{Buffer, Opus}
+
+  def_options self_delimit?: [
+                spec: boolean(),
+                default: false,
+                description: """
+                If true, will mutate the output to self-delimit the Opus input.
 
-  def_input_pad :input,
-    demand_unit: :buffers,
-    caps: [{Opus, self_delimiting?: true}, {RemoteStream, content_format: one_of([Opus, nil])}]
+                See https://tools.ietf.org/html/rfc6716#appendix-B for details.
+                """
+              ]
 
-  def_output_pad :output, caps: {Opus, self_delimiting?: false}
+  def_input_pad :input, demand_unit: :buffers, caps: :any
+  def_output_pad :output, caps: Opus
 
   @impl true
-  def handle_init(_opts) do
-    {:ok, %{partial_packet: <<>>, timestamp: 0}}
+  def handle_init(%__MODULE__{} = options) do
+    {:ok, options |> Map.from_struct()}
   end
 
   @impl true
-  def handle_caps(:input, _caps, _ctx, state) do
-    {:ok, state}
+  def handle_demand(:output, bufs, :buffers, _ctx, state) do
+    {{:ok, demand: {:input, bufs}}, state}
   end
 
   @impl true
-  def handle_demand(:output, _size, _unit, _ctx, state) do
-    {{:ok, demand: {:input, 1}}, state}
+  def handle_process(:input, %Buffer{payload: data}, _ctx, state) do
+    with {:ok, {configuration_number, stereo_flag, frame_packing}} <- parse_toc_byte(data),
+         {:ok, {_mode, _bandwidth, frame_size}} <- parse_configuration(configuration_number),
+         {:ok, channels} <- parse_channels(stereo_flag),
+         {:ok, {frame_lengths, header_length}} <- parse_frame_lengths(frame_packing, data),
+         {:ok, data} <- self_delimit(data, frame_lengths, header_length, state) do
+      caps = %Opus{
+        channels: channels,
+        self_delimiting?: state.self_delimit?
+      }
+
+      buffer = %Buffer{
+        payload: data,
+        metadata: %{
+          frame_lengths: frame_lengths,
+          frame_size: frame_size
+        }
+      }
+
+      {{:ok, caps: {:output, caps}, buffer: {:output, buffer}}, state}
+    else
+      {:error, reason} -> {{:error, reason}, state}
```

Comment:
i guess that's true. do you think it's bad to include the `else` just in case something changes later? i do like the `with` part for making the data flow clear. curious for your opinion here

Reply:
Having any code 'in case something changes later' is a bad practice IMHO (maybe unless you're sure this is going to change right away - then I'd add TODO/FIXME though). For me reading this code it didn't make anything clear, but only more confusing, as I expect the code to be there because it's needed. Moreover, it may happen that this code will never be used, while it has to be maintained. Regarding errors, you should either specify them and know exactly what errors can be returned, or `raise` right away. Returning errors only makes sense when someone is going to handle them - otherwise, it only introduces a mess. Membrane doesn't always follow this rule though, because it took me some time to realize that :P

Reply:
removed the `else`. I really do like the `with` part for showing data flow, as opposed to a series of variable sets. what do you think?

Reply:
I cannot agree, unfortunately. `with` suggests that something can fail and we're going to recover from that, which is not the case here. Moreover, now if something actually doesn't match, it gets silently returned, what may lead to a cryptic error. I understand you like `with`, but aesthetic issues shouldn't affect the code behaviour. That said, I think you're going to need that `with` if you handle self-delimiting input, as you're not guaranteed to get exactly one packet per buffer then. BTW you may want to check out my extension to with https://hexdocs.pm/bunch/Bunch.html#withl/2 :P